File: SQL/Select/SQLSelectTables.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLSelectTables
        Implements Collections.Generic.IEnumerable(Of SQLSelectTableBase)

        Private pobjTables As New Collections.Generic.List(Of SQLSelectTableBase)
        Private pobjJoins As SQLSelectTableJoins = New SQLSelectTableJoins

        Public Sub New()

            MyBase.New()

        End Sub

        Public Function Add() As SQLSelectTable

            Return Add(String.Empty)

        End Function

        Public Function Add(ByVal strTableName As String) As SQLSelectTable

            Dim objTable As SQLSelectTable = New SQLSelectTable(strTableName)

            pobjTables.Add(objTable)

            Return objTable

        End Function

        Public Function Add(ByVal strTableName As String, ByVal strAlias As String) As SQLSelectTable

            Dim objTable As SQLSelectTable = New SQLSelectTable(strTableName, strAlias)

            pobjTables.Add(objTable)

            Return objTable

        End Function

        Public Sub Add(ByVal objTable As SQL.SQLSelectTable)

            pobjTables.Add(objTable)

        End Sub

        Public Sub Add(ByVal objTable As SQL.SQLSelectTableFromSelect)

            pobjTables.Add(objTable)

        End Sub

        Default Public ReadOnly Property Item(ByVal strTableName As String) As SQLSelectTableBase
            Get

                Return Me(TableNameIndex(strTableName))

            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal intIndex As Integer) As SQLSelectTableBase
            Get

                Return pobjTables.Item(intIndex)

            End Get
        End Property

        Public ReadOnly Property Count() As Integer
            Get

                Return pobjTables.Count()

            End Get
        End Property

        Public Property Joins() As SQLSelectTableJoins
            Get

                Return pobjJoins

            End Get

            Set(ByVal Value As SQLSelectTableJoins)

                pobjJoins = Value

            End Set
        End Property

        Public Function Exists(ByVal strTableName As String) As Boolean

            Return TableNameIndex(strTableName) >= 0

        End Function

        Public Sub Delete(ByRef objTable As SQLSelectTable)

            If pobjTables.Contains(objTable) Then
                Throw New IndexOutOfRangeException
            End If

            pobjTables.Remove(objTable)
            objTable = Nothing

        End Sub

        Friend ReadOnly Property SQL(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = String.Empty
                Dim bAddTable As Boolean

                For intIndex As Integer = 0 To Me.Count - 1
                    'Include the table if it's not being used in a join
                    If pobjJoins Is Nothing Then
                        bAddTable = True
                    ElseIf Not pobjJoins.Exists(Me.Item(intIndex)) Then
                        bAddTable = True
                    Else
                        bAddTable = False
                    End If

                    If bAddTable Then
                        strSQL &= Me.Item(intIndex).SQL(eConnectionType)
                        If intIndex <> Me.Count - 1 Then
                            strSQL &= ", "
                        End If
                    End If
                Next

                If Not pobjJoins Is Nothing Then
                    If pobjJoins.SQL(eConnectionType) <> String.Empty And strSQL <> String.Empty Then
                        strSQL &= " "
                    End If
                    strSQL &= pobjJoins.SQL(eConnectionType)
                End If

                Return strSQL

            End Get
        End Property

        Private Function TableNameIndex(ByVal strTableName As String) As Integer

            Dim objTable As SQLSelectTable

            strTableName = strTableName.Trim()

            For intIndex As Integer = 0 To Me.Count - 1
                objTable = DirectCast(pobjTables.Item(intIndex), SQLSelectTable)
                If String.Compare(strTableName, objTable.Name, ignoreCase:=True) = 0 Then
                    Return intIndex
                End If
            Next

            Return -1

        End Function

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator

            Return pobjTables.GetEnumerator

        End Function

        Private Function GetEnumerator1() As System.Collections.Generic.IEnumerator(Of SQLSelectTableBase) Implements System.Collections.Generic.IEnumerable(Of SQLSelectTableBase).GetEnumerator

            Return pobjTables.GetEnumerator

        End Function

    End Class

End Namespace

File: SQL/Select/SQLSelectTables.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Collections.Generic;
using System.Linq;

namespace DatabaseObjects.SQL
{
	public class SQLSelectTables : IEnumerable<SQLSelectTableBase>
	{
		private SQLSelect pobjParent;
		private List<SQLSelectTableBase> pobjTables = new List<SQLSelectTableBase>();
		private SQLSelectTableJoins pobjJoins = new SQLSelectTableJoins();
			
		public SQLSelectTables()
		{
		}
			
		internal SQLSelectTables(SQLSelect objParent)
		{
			pobjParent = objParent;
		}
			
		/// <summary>
		/// Returns the associated parent SQLSelect object associated with this table collection.
		/// Returns nothing if the object is not assigned to an SQLSelect object.
		/// </summary>
		public SQLSelect Parent
		{
			get
			{
				return pobjParent;
			}
				
			internal set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				pobjParent = value;
			}
		}
			
		public SQLSelectTable Add()
		{
			return Add(string.Empty);
		}
			
		public SQLSelectTable Add(string strTableName)
		{
			SQLSelectTable objTable = new SQLSelectTable(strTableName);
				
			pobjTables.Add(objTable);
				
			return objTable;
		}
			
		public SQLSelectTable Add(string strTableName, string strAlias)
		{
			SQLSelectTable objTable = new SQLSelectTable(strTableName, strAlias);
				
			pobjTables.Add(objTable);
				
			return objTable;
		}
			
		public void Add(SQL.SQLSelectTable objTable)
		{
			pobjTables.Add(objTable);
		}
			
		public void Add(SQL.SQLSelectTableFromSelect objTable)
		{
			pobjTables.Add(objTable);
		}
			
		public SQLSelectTableBase this[string strTableName]
		{
			get
			{
				if (!Exists(strTableName))
					throw new ArgumentException(strTableName + " does not exist");

				return pobjTables.Single(table => Equals(table, strTableName));
			}
		}
			
		public SQLSelectTableBase this[int intIndex]
		{
			get
			{
				return pobjTables[intIndex];
			}
		}
			
		public int Count
		{
			get
			{
				return pobjTables.Count;
			}
		}
			
		public SQLSelectTableJoins Joins
		{
			get
			{
				return pobjJoins;
			}
				
			set
			{
				pobjJoins = value;
			}
		}
			
		public bool Exists(string strTableName)
		{
			return pobjTables.SingleOrDefault(table => Equals(table, strTableName)) != null;
		}
			
		public void Delete(ref SQLSelectTable objTable)
		{
			if (pobjTables.Contains(objTable))
				throw new IndexOutOfRangeException();
				
			pobjTables.Remove(objTable);
			objTable = null;
		}
						
		private bool Equals(SQLSelectTable table, string strTableName)
		{
			return table.Name.Equals(strTableName, StringComparison.InvariantCultureIgnoreCase);
		}
			
		public System.Collections.IEnumerator GetEnumerator()
		{
			return pobjTables.GetEnumerator();
		}
			
		IEnumerator<SQLSelectTableBase> IEnumerable<SQLSelectTableBase>.GetEnumerator()
	    {
			return pobjTables.GetEnumerator();
	    }

	    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
	    {
			return pobjTables.GetEnumerator();
	    }
	}
}